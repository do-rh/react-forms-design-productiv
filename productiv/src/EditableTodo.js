import React, { useState } from "react";
import Todo from "./Todo";
import TodoForm from "./TodoForm";

/** Show editable todo item.
 *
 * Props
 * - todo
 * - update(): fn to call to update a todo
 * - remove(): fn to call to remove a todo
 *
 * EditableTodoList -> EditableTodo -> { Todo, TodoForm }
 */

function EditableTodo({ todo, update, remove }) {
  const [isEditing, setEdit] = useState(false);
  const { title, description, priority } = todo;

  /** Toggle if this is being edited */
  function toggleEdit() {
    setEdit(isEditing => !isEditing);
  }

  /** Call remove fn passed to this. */
  function handleDelete() {
    remove(todo.id);
  }

  /** Edit form saved; toggle isEditing and update in ancestor. */
  function handleSave(formData) {
    //call update() fn in editableToDoList
    update(formData);
    toggleEdit();
  }


  return (
    <div className="EditableTodo">
      {isEditing
        ? <TodoForm initialFormData={todo} handleSave={handleSave} />
        : <div className="mb-3">
          <Todo title={title} description={description} priority={priority} />
          <div className="float-right text-sm-right">
            <button
              className="EditableTodo-toggle btn-link btn btn-sm"
              onClick={toggleEdit}>
              Edit
            </button>
            <button
              className="EditableTodo-delBtn btn-link btn btn-sm text-danger"
              onClick={handleDelete}>
              Del
            </button>
          </div>
        </div>
      }
    </div>
  );
}

export default EditableTodo;