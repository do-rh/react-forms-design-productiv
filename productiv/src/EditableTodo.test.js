import { render, fireEvent } from "@testing-library/react";
import EditableTodo from "./EditableTodo";

const TEST_TODO = {
    title: "test title",
    description: "meow meow meow",
    priority: 1
};

const update = jest.fn();
const remove = jest.fn();

it("renders without crashing", function () {
    render(<EditableTodo todo={TEST_TODO} remove={remove} update={update} />);
});

it("matches snapshot", function () {
    const { container, debug } = render(<EditableTodo todo={TEST_TODO} remove={remove} update={update} />);
    expect(container).toMatchSnapshot();
});

it("renders edit and delete buttons", function () {
    const { container, debug } = render(<EditableTodo todo={TEST_TODO} remove={remove} update={update} />);
    expect(container).toContainHTML(`<button
        className="EditableTodo-toggle btn-link btn btn-sm"`);
    expect(container).toContainHTML(`<button
        className="EditableTodo-delBtn btn-link btn btn-sm text-danger`);
});

xit("successfully calls update when clicked", function () {
    const { container, debug } = render(<EditableTodo todo={TEST_TODO} remove={remove} update={update} />);
    // const updateBtn = document.getElementsByClassName("EditableTodo-toggle btn-link btn btn-sm")[0];
    // fireEvent.click(updateBtn);
    // console.log(updateBtn)
    // expect(update).toBeCalledTimes(1);
});

it("successfully calls delete function (from prop) when clicked", function () {
    const { container, debug } = render(<EditableTodo todo={TEST_TODO} remove={remove} update={update} />);
    const deleteBtn = document.getElementsByClassName("EditableTodo-delBtn")[0];
    fireEvent.click(deleteBtn);
    expect(remove).toBeCalledTimes(1);
});


// it("shows user msg if todo list is empty", function () {
//     const { container, debug } = render(<EditableTodo todos={[]} remove={remove} update={update} />);
//     expect(container).toContainHTML('<span className="text-muted">You have no todos.</span>');
//     expect(container).not.toContainHTML('className="EditableTodo-toggle btn-link btn btn-sm"');
// });


it("unittest for toggleEdit", function () {
    let isEditing = true;
    toggleEdit();
    expect(isEditing).toBe(false);

    isEditing = false;
    toggleEdit();
    expect(isEditing).toBe(true);
});