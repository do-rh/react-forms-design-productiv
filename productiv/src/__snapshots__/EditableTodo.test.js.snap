// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="Todo"
      >
        <div>
          <b>
            test title
          </b>
           
          <small>
            Priority: 
            1
          </small>
        </div>
        <div>
          <small>
            meow meow meow
          </small>
        </div>
      </div>
      <div
        class="float-right text-sm-right"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        >
          Del
        </button>
      </div>
    </div>
  </div>
</div>
`;
